### PR TITLE
End-to-end tests

### PR DESCRIPTION
**Pull request by nmattia**
Add very basic end-to-end, screenshot tests that use the locally running replica (and hence https://github.com/dfinity/nns-dapp/pull/366). The screenshots are committed to the PRs that trigger the CI runs (see https://github.com/dfinity/nns-dapp/pull/367/commits/ae9a7c36b5a1b3eb2c083edc5a7df1f5e26c56ae). Having the CI upload the screenshots is really useful: if a PR changes a page's style this will be clear on the PR and the changes can be inspected easily by the reviewers.

---

Uses https://github.com/dfinity/nns-dapp/pull/366, which will need to be merged first.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -88,15 +89,67 @@ jobs:
...
           dfx canister --no-wallet create --all
+
+      - name: Deploy
+        run: |
+          DEPLOY_ENV=local dfx deploy --no-wallet --network local
+
+      - name: Run proxy
+        working-directory: proxy
+        run: |
+          npm ci
```
Comment (bitdivine) [inline .github/workflows/build.yml:104]: This shows one high severity vulnerability.  I thought we had `npm audit` and `cargo audit` in CI.  They should have caught this.  Just bumping versions fixes this though.  No big deal.

I checked.  We don't audit this code periodically.  We need to.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -88,15 +89,67 @@ jobs:
...
+          name: proxy-log
+          path: proxy/proxy.log
+
+      - name: Commit screenshots
+        uses: EndBug/add-and-commit@v7.4.0
+        with:
+          add: e2e-tests/screenshots
+          author_name: Screenshot Committer
+          author_email: "<nobody@example.com>"
+          message: "Update selenium screenshots"
```
Comment (bitdivine) [inline .github/workflows/build.yml:145]: Does webdriver use selenium?

File: e2e-tests/package.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "e2e-tests",
+    "version": "1.0.0",
+    "private": true,
+    "scripts": {
+        "wdio": "wdio run ./wdio.conf.ts"
```
Comment (bitdivine) [inline e2e-tests/package.json:6]: Does wdio support virtual webauthn?  I thought someone tried and couldn't make it work, hence selenium.

This looks as if a virtual webauthn *should* work with wdio but lacks examples: https://github.com/w3c/webauthn/pull/1256/files

File: e2e-tests/wdio.conf.ts [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    // },
+    /**
+     * Gets executed before test execution begins. At this point you can access to all global
+     * variables like `browser`. It is the perfect place to define custom commands.
+     * @param {Array.<Object>} capabilities list of capabilities details
+     * @param {Array.<String>} specs        List of spec file paths that are to be run
+     * @param {Object}         browser      instance of created browser/device session
+     */
+    // before: function (capabilities, specs) {
+    // },
```
Comment (bitdivine) [inline e2e-tests/wdio.conf.ts:244]: I suppose these are left in intentionally as they are useful if we want to fill them in later?

**Review by bitdivine — COMMENTED** [review]
# Feedback
Generally happy with this but there is one potential blocker and one issue that could reduce the scope of utility:
- [ ] I would like confirmation that webauthn works for us in this context.  Do you have a proof of concept to show that it does?
- [ ] How does wdio "mechanical turk" operations?  In Selenium I can use a real hardware wallet by letting jest run tests that pause and prompt me when it needs me to press some buttons on the hardware wallet.  One can of course mock but that just verifies that the code works if the hardware wallet behaves as we think it does, so one of the biggest danger points is not covered.

# Testing
I single stepped through the CI code and rushed through the manual instructions, less carefully.  Both works, assuming the modifications mentioned in the comments:
- Use `npm ci` rather than `npm install` as the latter can change the set of installed packages whereas `npm ci` just gets what is in the lock file, and is faster to boot.
- `npm run install-webdrivers` is not defined and does not seem to be needed.

File: e2e-tests/wdio.conf.ts (hunk 4 above)
Comment (nmattia) [reply e2e-tests/wdio.conf.ts:244]: Nope, overlooked. Thanks!

File: e2e-tests/package.json (hunk 3 above)
Comment (nmattia) [reply e2e-tests/package.json:6]: Jaein; can you give examples of why we'd want to support virtual webauthn?

File: .github/workflows/build.yml (hunk 2 above)
Comment (nmattia) [reply .github/workflows/build.yml:145]: Copy pasta, will remove

File: .github/workflows/build.yml (hunk 1 above)
Comment (nmattia) [reply .github/workflows/build.yml:104]: Gotcha! Will run locally. Until then, not really related to this PR so will resolve as is.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -88,15 +89,67 @@ jobs:
...
+      - name: Run proxy
+        working-directory: proxy
+        run: |
+          npm ci
+          # TODO: source the params from dfx.json and canister_ids.json
+          # https://dfinity.atlassian.net/browse/L2-200
+          npm run start -- \
+            --replica-host http://localhost:8080 \
+            rwlgt-iiaaa-aaaaa-aaaaa-cai:8086 \
+            > proxy.log &
```
[the remark this answers is not in the document]
Comment (nmattia) [reply .github/workflows/build.yml:110]: Agreed, got a task for that: https://dfinity.atlassian.net/browse/L2-200

Will address in a separate PR because there are some corner cases I want to look at first

File: e2e-tests/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+npm install
+npm run start --replica-host ...
+```
+
+Install all the dependencies for the test suite, and run the tests:
+
+```
+cd e2e-tests
+npm install
+npm run install-webdrivers
```
[the remark this answers is not in the document]
Comment (nmattia) [reply e2e-tests/README.md:44]: Left over from an early iteration, will remove

File: e2e-tests/README.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+# in another shell, deploy the canister
+DEPLOY_ENV=nobuild dfx deploy --no-wallet --network local
+```
+
+
+Install all the dependencies for the proxy, and run the proxy:
+
+``` bash
+cd proxy
+npm install
```
[the remark this answers is not in the document]
Comment (nmattia) [reply e2e-tests/README.md:35]: Wops, my bad!

File: e2e-tests/README.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+npm run start --replica-host ...
+```
+
+Install all the dependencies for the test suite, and run the tests:
+
+```
+cd e2e-tests
+npm install
```
[the remark this answers is not in the document]
Comment (nmattia) [reply e2e-tests/README.md:43]: Wops, my bad!

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -88,15 +89,67 @@ jobs:
...
+
+      - name: prepare and run the test suite
+        working-directory: e2e-tests
+        run: |
+          npm install
```
[the remark this answers is not in the document]
Comment (nmattia) [reply .github/workflows/build.yml]: Fixed!

File: e2e-tests/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (nmattia) [reply e2e-tests/README.md]: Fixed!

File: e2e-tests/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (nmattia) [reply e2e-tests/README.md]: Fixed!

File: e2e-tests/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (nmattia) [reply e2e-tests/README.md]: Removed!

File: .github/workflows/build.yml (hunk 2 above)
Comment (nmattia) [reply .github/workflows/build.yml:140]: Removed!

File: e2e-tests/wdio.conf.ts (hunk 4 above)
Comment (nmattia) [reply e2e-tests/wdio.conf.ts]: Removed!

File: e2e-tests/README.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+# in the repo root
+DEPLOY_ENV=nobuild dfx deploy --no-wallet --network local
+```
+
+Install all the dependencies for the test suite, and run the tests:
+
+```bash
+# in proxy/
+npm ci
+npm run build
```
Comment (nmattia) [inline e2e-tests/README.md:33]: @peterpeterparker is there any way we could avoid having to `npm run build` in `proxy/` first? I tried adding `npm build` as both the `preinstall` and `install` scripts of `proxy` but it doesn't quite do what we need

File: e2e-tests/package.json (hunk 3 above)
Comment (bitdivine) [reply e2e-tests/package.json]: Any testing that involves logging in.

Comment (nmattia) [reply e2e-tests/package.json]: clarified offline; will try to avoid the web authenticator as much as possible, but basically we'll cross that bridge when we get there

File: e2e-tests/README.md (hunk 10 above)
Comment (nmattia) [reply e2e-tests/README.md:33]: @peterpeterparker replied offline, he said it's ok.

**Review by bitdivine — APPROVED** [review]
